### PR TITLE
Cookie sharing topic update

### LINKED ISSUE #969: update "Sharing cookies between applications"
https://docs.asp.net/en/latest/security/data-protection/compatibility/cookie-sharing.html

Needs update.

### LINKED ISSUE #1812: Add a link to a working sample to the cookie sharing doc
https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/compatibility/cookie-sharing

I think we might be able to link to https://github.com/aspnet/Identity/tree/master/samples/Interop4.5

### LINKED ISSUE #3776: 2.0 auth update security/data-protection/compatibility/cookie-sharing
https://github.com/aspnet/Identity/issues/1321

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #3776
Fixes #1812 
Fixes #969
Addresses #3777 

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/security/data-protection/compatibility/cookie-sharing?branch=pr-en-us-4981&tabs=aspnetcore2x)

### CODE REVIEW

**Comment by HaoK** [comment]
Well not really since auth options changed to named options rather than instances living in identity options...  so it'd look something more like:

```
services.ConfigureApplicationCookie(o => o.DataProtectionProvider = 
        DataProtectionProvider.Create(new DirectoryInfo(@"c:\shared-auth-ticket-keys\"));
```

**Comment by guardrex** [comment]
> Well not really since auth options changed to named options

Since that change might be present in other parts of the topic, I'll sweep the whole thing.

**Comment by HaoK** [comment]
This whole scenario is pretty fragile in general, make sure you try the side by side interop since its very easy for it to be broken

**Comment by guardrex** [comment]
Sounds good. I'll run the whole topic. If I'm setting up something locally to test the guidance in this topic and it looks like it would make a good sample, I'll put it up in a repo so we can look at it and decide if it should accompany the topic.

**Comment by guardrex** [comment]
@HaoK I'm working on Core app sharing first.

The trick to making two cookie auth apps (without Identity) share was just to add ...
```csharp
options.Cookie.SameSite = SameSiteMode.None;
```
... resulting in ...
```csharp
services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
    .AddCookie(options =>
    {
        options.DataProtectionProvider = DataProtectionProvider.Create(
            new DirectoryInfo(
                @"C:\Users\XXXXX\Documents\GitHub\CookieSharingSample\shared-auth-ticket-keys\"));
        options.Cookie.SameSite = SameSiteMode.None;
    });
```
They share cookies. :+1:

Attempting to share that cookie with a third app that uses Identity is failing. The app can see the cookie, but it isn't using it. I tried with and without setting `TicketDataFormat` ...
```csharp
services.AddIdentity<ApplicationUser, IdentityRole>()
    .AddEntityFrameworkStores<ApplicationDbContext>()
    .AddDefaultTokenProviders();

services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
    .AddCookie(options =>
    {
        var protectionProvider = DataProtectionProvider.Create(
            new DirectoryInfo(
                @"C:\Users\XXXXX\Documents\GitHub\CookieSharingSample\shared-auth-ticket-keys\"));
        options.DataProtectionProvider = protectionProvider;
        /*
        options.TicketDataFormat = 
            new TicketDataFormat(
                protectionProvider.CreateProtector(
                    "Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware", 
                    "Cookies", 
                    "v2"));
        */
        options.Cookie.SameSite = SameSiteMode.None;
    });
```
I think I'm missing a config setting.

**Comment by guardrex** [comment]
@HaoK kk ... After some rubber ducking today, I have a setup that seems to work across cookie auth without identity and cookie auth with identity. Here are the critical components:

### Cookie auth without Identity app
*I'll call this the non-Identity app. I made two of these so that I could test the non-Identity-to-non-Identity case.*

*Startup.cs*
```csharp
services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
    .AddCookie(CookieAuthenticationDefaults.AuthenticationScheme, options =>
    {
        options.Cookie.Name = ".AspNetCore.Identity.Application";
        options.Cookie.SameSite = SameSiteMode.None;
        options.DataProtectionProvider = DataProtectionProvider.Create(
            new DirectoryInfo(
                @"C:\Users\XXXXX\Documents\GitHub\CookieSharingSample\shared-auth-ticket-keys\"));
    });
```
The important settings here are matching the cookie name to the cookie name used by the Identity app. Set `SameSite` to `None`. Setup a common DP key location across apps. I'm not clear on the double-setting of `CookieAuthenticationDefaults.AuthenticationScheme` yet. I need to run another couple of tests on that.

Note that I tried (repeatedly 😅) to get the Identity app to use the default `.AspNetCore.Cookie` cookie name by not setting it or setting it explicitly across the test apps. No 🎲🎲. Since the Identity app seems so hell bent on using the `.AspNetCore.Identity.Application` cookie name, I flipped the script on it and made my two non-Identity apps use that name, which seems to have worked.

*Login.cshtml.cs*
```csharp
var claims = new List<Claim>
{
    ...
};

var claimsIdentity = new ClaimsIdentity(claims, IdentityConstants.ApplicationScheme);

var authProperties = new AuthenticationProperties() { };

await HttpContext.SignInAsync(CookieAuthenticationDefaults.AuthenticationScheme, 
    new ClaimsPrincipal(claimsIdentity), authProperties);
```
The important point here is setting the auth type to `IdentityConstants.ApplicationScheme` for the principal's claims identity. If not set, then the Razor view call to `IsSignedIn` ([reference source](https://github.com/aspnet/Identity/blob/eb3ff7fc32dbfff65a1ba6dfdca16487e0f6fc41/src/Microsoft.AspNetCore.Identity/SignInManager.cs#L126-L127)) won't function properly ... the UI won't display Logoff button properly in the Identity app when the user authenticated from the non-Identity app.

### Cookie auth with Identity app
*I'm calling this the Identity app.*

*Startup.cs*
```csharp
services.AddIdentity<ApplicationUser, IdentityRole>()
    .AddEntityFrameworkStores<ApplicationDbContext>()
    .AddDefaultTokenProviders();

services.ConfigureApplicationCookie(options => {
    options.Cookie.SameSite = SameSiteMode.None;

    var protectionProvider = DataProtectionProvider.Create(
        new DirectoryInfo(
            @"C:\Users\XXXXX\Documents\GitHub\CookieSharingSample\shared-auth-ticket-keys\"));

    options.DataProtectionProvider = protectionProvider;

    options.TicketDataFormat = 
            new TicketDataFormat(
                protectionProvider.CreateProtector(
                    "Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware", 
                    "Cookies", 
                    "v2"));
});
```
Important settings: `SameSite` to `None`, the common DP key location, and setting the `TicketDataFormat`.

*AccountController.cs*
Don't specify a scheme for the `SignOutAsync` calls; otherwise, sign out from the Identity app doesn't occur properly. I presume it's using the "Cookies" scheme when it isn't specified in the call.
```csharp
await HttpContext.SignOutAsync();
```
This setup seems to work across Core apps.

Next round (Thursday) is to get these three apps talk'in to an old skool .NET Framework 4.6.1 app.

**Comment by guardrex** [comment]
### .NET Framework 4.6.1
* Install `Microsoft.Owin.Security.Interop`
* Mod `UseCookieAuthentication` as shown below ...

```csharp
app.UseCookieAuthentication(new CookieAuthenticationOptions
{
    AuthenticationType = DefaultAuthenticationTypes.ApplicationCookie,
    CookieName = ".AspNetCore.Identity.Application",
    LoginPath = new PathString("/Account/Login"),
    Provider = new CookieAuthenticationProvider
    {
        // Enables the application to validate the security stamp when the user logs in.
        // This is a security feature which is used when you change a password or add an external login to your account.  
        OnValidateIdentity = SecurityStampValidator.OnValidateIdentity<ApplicationUserManager, ApplicationUser>(
            validateInterval: TimeSpan.FromMinutes(30),
            regenerateIdentity: (manager, user) => user.GenerateUserIdentityAsync(manager))
    },
    TicketDataFormat = new AspNetTicketDataFormat(
        new DataProtectorShim(
            DataProtectionProvider.Create(new DirectoryInfo(@"C:\Users\XXXXX\Documents\GitHub\CookieSharingSample\shared-auth-ticket-keys\"))
         .CreateProtector("Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware",
            "Cookies", "v2"))),
                CookieManager = new ChunkingCookieManager()
});

// If not setting http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier and 
// http://schemas.microsoft.com/accesscontrolservice/2010/07/claims/identityprovider then set
// UniqueClaimTypeIdentifier to a claim that distinguishes unique users.
System.Web.Helpers.AntiForgeryConfig.UniqueClaimTypeIdentifier = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name";
```
* Set the `CookieName` to match.
* Configure the `TicketDataFormat` with the path to the shared DP key location.
* Set `UniqueClaimTypeIdentifier` if `nameidentifier` **and** `identityprovider` aren't provided by the ClaimsIdentity.

**Comment by HaoK** [comment]
Looks reasonable, the only new thing seems to be the `options.Cookie.SameSite = SameSiteMode.None;` which wasn't required before

**Comment by HaoK** [comment]
@blowdart @Tratcher do any of these changes look unexpected to you?

**Comment by guardrex** [comment]
@HaoK I'll test that again, but without it, it was a :boom: outcome. May have been because of POSTs (e.g., to Logoff actions)?? (maybe ... guessing here ... I'll take a closer look)

**Comment by HaoK** [comment]
I don't think you need to double check that, its probably something that was changed fairly recently

**Comment by guardrex** [comment]
The cookie name: My Identity Core 2.0 test app absolutely insists on using the name `.AspNetCore.Identity.Application`. I might be setting it incorrectly. I was trying to use `options.Cookie.Name` to make it use `.AspNetCore.Cookie`, but it didn't care to pay attention to my setting that. This is why I flipped and decided to force the other test apps to use `.AspNetCore.Identity.Application` ... I got sick of fighting it and let that app win the battle in order to win the war.

I also need to check on the double-setting of ... `CookieAuthenticationDefaults.AuthenticationScheme` ...
```csharp
services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
    .AddCookie(CookieAuthenticationDefaults.AuthenticationScheme, options =>
```
It might be required in both or either spots. I'd like to work that out.

**Comment by HaoK** [comment]
Yes identity core 2.0 removed the ability to change the scheme names for the identity cookies, but you probably should be able to tweak the cookie name though.  Can you file an issue in Identity with the stuff you tried and didn't work?  I can take a look at that later to see if there's any issues in the code.

**Comment by HaoK** [comment]
You can probably simplify to:

```
services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
    .AddCookie(options =>
```

Since that should be the default value for AddCookie

**Comment by guardrex** [comment]
Confirmed ...
```csharp
services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
    .AddCookie(options =>
```
:+1:
... and 😠 I should've guessed that `.AspNetCore.Identity.Application` was going to make a liar out of me. I could have swore it was insisting on that before. I was deleting the cookie manually. Anyway ... I'm going to flip these other apps over to using the default cookie name now and make sure all is good.

**Comment by HaoK** [comment]
Its not you, this gave both me and @blowdart a lot of grief when we were trying to set these up.  The steps change slightly over time unfortunately :(

**Comment by guardrex** [comment]
@Tratcher It doesn't. 😓 There was a step where I had some whacko config that resulted in things not working for POSTs (IIRC) when it was left at the default `Lax`. Later after the dust settled on configuration, I went back and checked, and it's fine with the default `Lax`, so I've removed those lines.

**Comment by guardrex** [comment]
Don't look yet ... I have a little more info to add, and the text is still a bit rough.

I'll ping tomorrow .... er ... *later today* .... when it's ready for :eyes:.

**Comment by guardrex** [comment]
I have a 1st draft up now.

* This includes a sample solution of three projects. 
* Since this shows three apps and the whole subject is about cookie sharing across more than just two apps, I go with "among" over "between," including in the title.

**Comment by Rick-Anderson** [comment]
@guardrex  what do we need to merge this?

**Comment by guardrex** [comment]
It needs review. I wish another dev/author would pull the sample (a collection of apps) down and run to confirm that they all play nicely together (i.e., share a cookie without blowing up). The topic changes need a look :eyes:, too.

**Comment by HaoK** [comment]
If you want, you could try to adding the interop sample apps to the https://github.com/aspnet/AuthSamples repo, those are built as part of CI builds so it would allow docs to point to something that's never going to be stale.

**Comment by HaoK** [comment]
Basically something like
- Cookies.Interop.AspNetIdentity
- Cookies.Interop.Razor
- Cookies.Interop.Mvc

And a short readme describing what you need to do to run them and demonstrate that they share cookeis

**Comment by HaoK** [comment]
The long term benefit of having these live here, is we can eventually add some functional tests that verify that cookie sharing works using these apps.

**Comment by Tratcher** [comment]
Except we'd only want the docs to reference release branches, not dev.

**Comment by guardrex** [comment]
A lot of time was spent* (and :beer: consumed! :grinning:) getting these apps sorted out. However, I totally agree that hardening docs to future changes is the way to go where possible.

*Well, just a few hours really.

The three that I have in there are:

* ASP.NET Core 2.0 Razor Pages app without using [ASP.NET Core Identity](xref:security/authentication/identity)
* ASP.NET Core 2.0 MVC app with ASP.NET Core Identity
* ASP.NET Framework 4.6.1 MVC app with ASP.NET Identity

I thought a with-Identity and without-Identity pair would be good to have.

Is the *Cookies.Interop.AspNetIdentity* project a netfx 4.6.x app?

**Comment by HaoK** [comment]
Yeah AspNetIdentity is trying to disambiguate calling out that its not AspNetCore

**Comment by HaoK** [comment]
Ok so better names might be.
- Cookies.Interop.MvcIdentity
- Cookies.Interop.RazorNoIdentity
- Cookies.Interop.AspNetIdentity2

**Comment by guardrex** [comment]
SGTM

> you could try to adding the interop sample apps to the https://github.com/aspnet/AuthSamples repo, those are built as part of CI builds so it would allow docs to point to something that's never going to be stale.

@Rick-Anderson Your call. As you know, the contract is days away from ending, and I have enough to keep me busy through early next week. I can take a stab at this in the last couple of days before the axe comes down. :skull: :smile:

**Comment by guardrex** [comment]
Just in case this happens, I have a concern: If the samps are over in *AuthSamples*, I'm not going to be able to display the code from the sample files in the topic. We'd be moving to hardcoding the code into the topic. That'll be unstable if the samples shift over time and an issue isn't opened over here to fix them up.

**Comment by guardrex** [comment]
@Rick-Anderson UPDATE

I put a copy of the sample (three cookie-sharing projects) from this PR up in the AuthSamples repo with some modifications for the setup over there. @HaoK suggested that they would work better with a common Identity dB. For this topic, the dev will only need to create the user in one project to sign into all three apps. It will be an example of sharing an Identity dB generally across apps (Core-to-Core and Core-to-NET). He's going to work that forward.

In the meantime, we should continue to review+merge this as it stands. I'll open a new issue to refactor the topic and kill the sample here when the work is done over there.

@HaoK, this still requires a technical review and signoff if ur available to do so. Except for the README updates I just made along the lines of what you suggested over in AuthSamples, I plan to leave namespaces and project names alone given this will be short-lived over here.

**Comment by scottaddie** [comment]
@HaoK Please review.

File: aspnetcore/security/data-protection/compatibility/cookie-sharing.md [hunk 1]
```diff
@@ -1,112 +1,184 @@
 ---
-title: Sharing cookies between applications
+title: Sharing cookies among applications
```
Comment (scottaddie) [inline aspnetcore/security/data-protection/compatibility/cookie-sharing.md]: applications --> apps

File: aspnetcore/security/data-protection/compatibility/cookie-sharing.md [hunk 2]
```diff
@@ -1,112 +1,184 @@
...
 author: rick-anderson
-description: This document explains how the data protection stack supports sharing of authentication cookies between ASP.NET 4.x and ASP.NET Core apps.
+description: Learn how to share authentication cookies among ASP.NET 4.x and ASP.NET Core apps.
 ms.author: riande
 manager: wpickett
-ms.date: 10/14/2016
+ms.custom: mvc
+ms.date: 12/18/2017
```
Comment (scottaddie) [inline aspnetcore/security/data-protection/compatibility/cookie-sharing.md]: Update the date

File: aspnetcore/security/data-protection/compatibility/cookie-sharing.md [hunk 3]
```diff
@@ -1,112 +1,184 @@
...
+  * A common [data protection key](xref:security/data-protection/implementation/key-management) storage location is used. The sample app uses a folder named *KeyRing* at the root of the solution to hold the data protection keys.
+  * [DataProtectionProvider.Create(System.IO.DirectoryInfo)](/dotnet/api/microsoft.aspnetcore.dataprotection.dataprotectionprovider.create?view=aspnetcore-2.0#Microsoft_AspNetCore_DataProtection_DataProtectionProvider_Create_System_IO_DirectoryInfo_) accepts a [DirectoryInfo](/dotnet/api/system.io.directoryinfo) for use with authentication cookies. The sample app provides the path of the *KeyRing* folder to `DirectoryInfo`.
+  * [DataProtectionProvider](/dotnet/api/microsoft.aspnetcore.dataprotection.dataprotectionprovider) requires the [Microsoft.AspNetCore.DataProtection.Extensions](https://www.nuget.org/packages/Microsoft.AspNetCore.DataProtection.Extensions/) NuGet package. The package is available to ASP.NET Core 2.0 apps when referencing the [Microsoft.AspNetCore.All](xref:fundamentals/metapackage) metapackage ([NuGet.org](https://www.nuget.org/packages/Microsoft.AspNetCore.All/)).
```
Comment (scottaddie) [inline aspnetcore/security/data-protection/compatibility/cookie-sharing.md]: Make it clear that the metapackage is intended for apps targeting .NET Core 2.0. Also call out explicitly that apps targeting .NET Framework must add this individual package.

**Comment by guardrex** [comment]
Thanks @scottaddie - Changes made.

**Comment by guardrex** [comment]
I'm OOF ATM, but IIRC this is RTM.

**Comment by guardrex** [comment]
This looks good ... I'll go ahead and shoot it in.

I opened an issue for the upcoming AuthSamples updates: https://github.com/aspnet/Docs/issues/5238

**Pull request merged**